contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -1511,12 +1511,28 @@ pltsql_strpos_non_determinstic(text *src_text, text *substr_text, Oid collid, in
 		     u16_pos != USEARCH_DONE;
 		     u16_pos = usearch_next(usearch, &status))
 		{
+			int32_t matched_idx = usearch_getMatchedStart(usearch);
+
 			if (U_FAILURE(status))
 				ereport(ERROR,
 					(errcode(ERRCODE_INTERNAL_ERROR),
 					 errmsg("failed to perform ICU search: %s",
 							u_errorName(status))));
 
+			/* ICU bug, When pattern start with a surrogate pair ICU usearch_next stops moving forward entering an infinite loop */
+			if (u16_pos == pos_prev_loop)
+			{
+				if (U16_IS_SURROGATE(src_uchar[matched_idx]) && U16_IS_SURROGATE(substr_uchar[0]) && matched_idx + 2 < src_ulen)
```

Comment:
Does the problem arises only when the first char is surrogate pair? I think better to use U16_IS_SURROGATE(substr_uchar[u16_pos]).

Reply:
done

Reply:
Yes, it only happens when substring to find start with a surrogate pair.
In this case ICU doesn't seem to set the offset for the next search correctly.